### DISCUSSION TITLE
gasket-driver can't be loaded on self-build kernel

### DISCUSSION THREAD

**Discussion by O-corn** [siderolabs/talos#7732] [Q&A] [answered]
When I use official installer image: ghcr.io/siderolabs/installer:v1.4.6 and extensions image: ghcr.io/siderolabs/gasket-driver:97aeba58-v1.4.6 to install Talos, gasket-driver can be loaded successfully.

**talosctl dmesg |grep gasket**
![image](https://github.com/siderolabs/talos/assets/44917389/1a6dff64-34b8-47cb-b3cf-02803250ff76)

**talosctl read /proc/modules**
![image](https://github.com/siderolabs/talos/assets/44917389/5b854d61-f975-4468-be2e-264a7005d2c9)

However, when I tried to use my self-build installer image and corresponding gasket-driver image to install Talos, gasket driver can't be loaded.

**talosctl dmesg |grep gasket**
![image](https://github.com/siderolabs/talos/assets/44917389/a315c72e-82e7-48bd-9d62-bf2f21e43139)

**talosctl read /proc/modules**
shows nothing

Is this issue related with the kernel config parameters. Or can you give some hints on it.

And If I want to insert an oot driver to talos, what should I do?  Can I follow the way https://github.com/siderolabs/pkgs/blob/main/gasket-driver/pkg.yaml does? Dose it work?

**Comment by smira** [comment, marked as answer]
Talos Linux kernel is using an ephemeral (discarded during the build) key to sign kernel modules, and it only allows loading modules signed with that key.

When you use official installer and extensions, they were built from the same kernel package, so the key matches.

When you use your own installer (kernel), the key is different.

So tl;dr is that if you build yourself, you need to make sure every kernel module is coming from the same build as the kernel itself.

**Reply by mi-bauer** [reply]
> So tl;dr is that if you build yourself, you need to make sure every kernel module is coming from the same build as the kernel itself.

Does it really need to be the same build (as in the process that builds the kernel)? Or would it be possible to create an extension that uses an kernel image that was built somewhere in the past? I assume that the extension would then of course only work with that old kernel. 

I am just confused, because that gasket driver seems to build its modules after the build of the kernel is done already. Thank you for any clarification. :)

**Reply by smira** [reply]
"Same build" in the sense that they share same buildkit cache if you want this to be more precise statement.

Talos has many out-of-tree kernel modules.
